src/tools/sfdk/dbus.cpp

```diff
@@ -72,24 +82,36 @@ class ConfiguredDeviceAdaptor : public AbstractAdaptor
     }
 
 public slots:
-    bool prepare(const QDBusMessage &message)
+    void prepare(const QDBusMessage &message)
     {
+        withDevice(message, [=](Device *device) {
+            RemoteProcess test;
+            test.setProgram("/bin/true");
+            test.setRunInTerminal(false);
+            if (!SdkManager::prepareForRunOnDevice(*device, &test)
+                    || test.exec() != EXIT_SUCCESS) {
+                replyError(message, tr("Failed to prepare device"));
+            }
+        });
+    }
+
+private:
+    template<typename Fn>
+    auto withDevice(const QDBusMessage &message, Fn fn) const
+        -> decltype(fn(std::declval<Device *>()))
```

Comment:
I do have to wonder if there's an easier way to do this. Maybe there isn't, but this just feels over complicated.

Reply:
What do you mean more exactly?

Reply:
Now I know what I mean, it took me a while to even realize it myself, as I had to start writing this comment in order to get my thoughts in order! :-D
The way I see this, the return value of withDevice() is only used by isEmulator(), which returns a bool. Using decltype and std::declval to deduce the return type feels a bit excessive when the the only used return values are bool or void. We could use e.g. std::optional for this. But then again, this is of course more future proof. So yeah, I don't have any real suggestions for improvements here, it's just trying to figure out what's happening here is not that easy. It's not really your fault but rather the C++ syntax which is not always as readable as I wish it was.